### PR TITLE
Generate FastPath for Array

### PR DESCRIPTION
**Pull request by kunalspathak**
Generate FastPath for `Array` ctor if it has single argument which is constant between 0 and 8 (inclusive).

Tests: In Progress. 
Benchmarks: See ~1.96% in Octane's Richard. Others, no change.

Fixes: #1271

### CODE REVIEW

**Comment by kunalspathak** [comment]
@rajatd , @LouisLaf - please review.

File: lib/Backend/Lower.cpp [hunk 1]
```diff
@@ -4932,15 +4934,47 @@ Lowerer::LowerNewScObjArray(IR::Instr *newObjInstr)
...
+        int32 length = linkSym->GetIntConstValue();
+
+        if (length >= 0 && length <= 8) // #2
         {
-            GenerateProfiledNewScObjArrayFastPath(newObjInstr, arrayInfo, weakFuncRef, (uint32)length);
+            AssertMsg(linkSym->IsArgSlotSym(), "Not an argSlot symbol...");
+            opndSrc = argInstr->GetSrc2();
+
+            // If more than 1 args, then opndSrc would be symOpnd
+            bool emitFastPath = !opndSrc->IsSymOpnd(); // #3
```
Comment (ThomsonTan) [inline lib/Backend/Lower.cpp]: >opndSrc [](start = 33, length = 7)

Could opndSrc be NULL?

Comment (kunalspathak) [inline lib/Backend/Lower.cpp]: No because `argInstr` is equivalent of `ArgOut_A` which always have associated `src2` that either points to previous `ArgOut_A` or `StartCall`.

File: lib/Backend/Lower.cpp [hunk 2]
```diff
@@ -4932,15 +4934,47 @@ Lowerer::LowerNewScObjArray(IR::Instr *newObjInstr)
...
+    Assert(opndSrc->IsSymOpnd());
+    StackSym *linkSym = opndSrc->AsSymOpnd()->m_sym->AsStackSym();
+    Assert(linkSym->IsSingleDef());
+    IR::Instr* argInstr = linkSym->GetInstrDef();
+    IR::Opnd *opndOfArrayCtor = argInstr->GetSrc1();
+    // Generate fast path only if it meets all the conditions:
+    // 1. 1st paramter is a constant
+    // 2. It is in range 0 <= parameter <= 8
+    // 3. It is the only parameter.
+    if (opndOfArrayCtor->IsIntConstOpnd() || opndOfArrayCtor->IsAddrOpnd()) // #1
```
Comment (satheeshravi) [inline lib/Backend/Lower.cpp]: >opndOfArrayCtor->IsAddrOpnd() [](start = 45, length = 29)

Why is this required ?

File: lib/Backend/Lower.cpp [hunk 3]
```diff
@@ -4932,15 +4934,47 @@ Lowerer::LowerNewScObjArray(IR::Instr *newObjInstr)
         Assert(weakFuncRef);
     }
 
-    IR::Opnd *opndSrc1 = newObjInstr->UnlinkSrc1();
-    if (opndSrc1->IsImmediateOpnd())
+    IR::LabelInstr *labelDone = IR::LabelInstr::New(Js::OpCode::Label, func);
+    IR::Opnd *opndSrc = newObjInstr->GetSrc2();
```
Comment (satheeshravi) [inline lib/Backend/Lower.cpp]: >opndSrc [](start = 14, length = 7)

nit: rename to linkOpnd

File: lib/Backend/Lower.cpp (hunk 1 above)
Comment (satheeshravi) [inline lib/Backend/Lower.cpp]: >!opndSrc->IsSymOpnd();  [](start = 32, length = 23)

Shouldn't you be looking at the slotNum(number of arguments passed) instead of this ? Won't second opnd always be SymOpnd ?

File: lib/Backend/Lower.cpp [hunk 4]
```diff
@@ -4932,15 +4934,47 @@ Lowerer::LowerNewScObjArray(IR::Instr *newObjInstr)
...
+    if (opndOfArrayCtor->IsIntConstOpnd() || opndOfArrayCtor->IsAddrOpnd()) // #1
     {
-        intptr_t length = opndSrc1->GetImmediateValue();
-        if (length >= 0 && length <= 8)
+        int32 length = linkSym->GetIntConstValue();
+
+        if (length >= 0 && length <= 8) // #2
         {
-            GenerateProfiledNewScObjArrayFastPath(newObjInstr, arrayInfo, weakFuncRef, (uint32)length);
+            AssertMsg(linkSym->IsArgSlotSym(), "Not an argSlot symbol...");
```
Comment (satheeshravi) [inline lib/Backend/Lower.cpp]: > AssertMsg(linkSym->IsArgSlotSym() [](start = 11, length = 34)

This assert and the below seems to be contradicting... did both of them pass ?

File: lib/Backend/Lower.cpp (hunk 2 above)
Comment (satheeshravi) [inline lib/Backend/Lower.cpp]: >(opndOfArrayCtor->IsIntConstOpnd() [](start = 7, length = 34)

since you are already doing this, you can directly call opndOfArrayCtor->GetImmediateValue ?

File: lib/Backend/Lower.cpp [hunk 5]
```diff
@@ -3855,8 +3859,9 @@ Lowerer::GenerateProfiledNewScObjArrayFastPath(IR::Instr *instr, Js::ArrayCallSi
     {
         GenerateArrayInfoIsNativeFloatAndNotIntArrayTest(instr, arrayInfo, helperLabel);
         Assert(Js::JavascriptNativeFloatArray::GetOffsetOfArrayFlags() + sizeof(uint16) == Js::JavascriptNativeFloatArray::GetOffsetOfArrayCallSiteIndex());
-        headOpnd = GenerateArrayAlloc<Js::JavascriptNativeFloatArray>(instr, &size, arrayInfo, &isZeroed);
+        headOpnd = GenerateArrayAlloc<Js::JavascriptNativeFloatArray>(instr, &size, arrayInfo, &isZeroed, containMissingItems);
 
+        GenerateMemInit(dstOpnd, Js::JavascriptNativeIntArray::GetOffsetOfArrayCallSiteIndex(), IR::IntConstOpnd::New(profileId, TyUint16, func, true), instr, isZeroed);
```
Comment (satheeshravi) [inline lib/Backend/Lower.cpp]: >JavascriptNativeIntArray [](start = 37, length = 24)

Should be JavascriptNactiveFloatArray

File: lib/Backend/Lower.cpp [hunk 6]
```diff
@@ -3833,16 +3835,18 @@ Lowerer::GenerateProfiledNewScObjArrayFastPath(IR::Instr *instr, Js::ArrayCallSi
     Func * func = this->m_func;
     IR::LabelInstr * helperLabel = IR::LabelInstr::New(Js::OpCode::Label, func, true);
     uint32 size = length;
+    bool containMissingItems = size != 0;
```
Comment (satheeshravi) [inline lib/Backend/Lower.cpp]: > size != 0 [](start = 30, length = 10)

nit: can you put this in a if condition.

File: lib/Backend/Lower.cpp [hunk 7]
```diff
@@ -3884,11 +3889,8 @@ Lowerer::GenerateProfiledNewScObjArrayFastPath(IR::Instr *instr, Js::ArrayCallSi
     }
 
     // Skip pass the helper call
-    IR::LabelInstr * doneLabel = IR::LabelInstr::New(Js::OpCode::Label, func);
```
Comment (satheeshravi) [inline lib/Backend/Lower.cpp]: > IR::LabelInstr * doneLabel = IR::LabelInstr::New(Js::OpCode::Label, func); [](start = 3, length = 75)

what is the motivation for moving this at the call site ? Is it any different ?

File: lib/Backend/Lower.cpp [hunk 8]
```diff
@@ -3823,7 +3825,7 @@ Lowerer::GenerateArrayAlloc(IR::Instr *instr, uint32 * psize, Js::ArrayCallSiteI
 }
 
 void
-Lowerer::GenerateProfiledNewScObjArrayFastPath(IR::Instr *instr, Js::ArrayCallSiteInfo * arrayInfo, RecyclerWeakReference<Js::FunctionBody> * weakFuncRef, uint32 length)
+Lowerer::GenerateProfiledNewScObjArrayFastPath(IR::Instr *instr, Js::ArrayCallSiteInfo * arrayInfo, RecyclerWeakReference<Js::FunctionBody> * weakFuncRef, uint32 length, IR::LabelInstr* labelDone)
```
Comment (satheeshravi) [inline lib/Backend/Lower.cpp]: >labelDone [](start = 186, length = 9)

do you need this ?

File: lib/Backend/Lower.cpp (hunk 1 above)
Comment (satheeshravi) [inline lib/Backend/Lower.cpp]: If you are doing this, you can consider adding this check in the outer if as a conjunction.

---
In reply to: [71630471](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71630471) [](ancestors = 71630471)

File: lib/Backend/Lower.cpp (hunk 5 above)
Comment (kunalspathak) [inline lib/Backend/Lower.cpp]: good catch. Copy-paste issue. :)

---
In reply to: [71650416](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71650416) [](ancestors = 71650416)

File: lib/Backend/Lower.cpp (hunk 6 above)
Comment (kunalspathak) [inline lib/Backend/Lower.cpp]: Sure, I can. Any reason for having it in if-condition vs. having it like this?

---
In reply to: [71650529](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71650529) [](ancestors = 71650529)

File: lib/Backend/Lower.cpp (hunk 1 above)
Comment (kunalspathak) [inline lib/Backend/Lower.cpp]: No. If opnd is from StartCall, it is of type RegOpnd. If it is from ArgsOut, it is of type SymOpnd.

---
In reply to: [71649040](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71649040) [](ancestors = 71649040)

File: lib/Backend/Lower.cpp (hunk 4 above)
Comment (kunalspathak) [inline lib/Backend/Lower.cpp]: Yes, this one ensures that `linkSym` is `ArgSlotSym` and the one below ensures that previous source's symbol is not `ArgSlotSym` i.e. it is for `StartCall`.

---
In reply to: [71649258](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71649258) [](ancestors = 71649258)

File: lib/Backend/Lower.cpp (hunk 2 above)
Comment (kunalspathak) [inline lib/Backend/Lower.cpp]: I was expecting the same, but surprisingly, opndOfArrayCtor is still of type `AddrOpnd` (may be because it is referring to ArgOut?). With that said, i think i should remove `opndOfArrayCtor->IsIntConstOpnd() ` condition.

---
In reply to: [71648857](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71648857) [](ancestors = 71648857)

Comment (kunalspathak) [inline lib/Backend/Lower.cpp]: Apparently opndOfArrayCtor is of type `OpndKindAddr` and not `OpndKindIntConst`.

---
In reply to: [71649822](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71649822) [](ancestors = 71649822)

File: lib/Backend/Lower.cpp (hunk 6 above)
Comment (ThomsonTan) [inline lib/Backend/Lower.cpp]: Or put parentheses around the compare expression?

---
In reply to: [71740566](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71740566) [](ancestors = 71740566,71650529)

File: lib/Backend/Lower.cpp (hunk 3 above)
Comment (kunalspathak) [inline lib/Backend/Lower.cpp]: sure.

---
In reply to: [71648999](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71648999) [](ancestors = 71648999)

File: lib/Backend/Lower.cpp (hunk 6 above)
Comment (kunalspathak) [inline lib/Backend/Lower.cpp]: Sure.

File: lib/Backend/Lower.cpp (hunk 1 above)
Comment (kunalspathak) [inline lib/Backend/Lower.cpp]: I can't do this without updating opndSrc with `argInstr->GetSrc2()`.

---
In reply to: [71650995](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71650995) [](ancestors = 71650995,71630471)

File: lib/Backend/Lower.cpp (hunk 2 above)
Comment (satheeshravi) [inline lib/Backend/Lower.cpp]: You can still call GetImmediateValue on the addr opnd, instead of this - in that way, you don't have to work around the fetching business.

---
In reply to: [71742831](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71742831) [](ancestors = 71742831,71649822)

File: lib/Backend/Lower.cpp (hunk 1 above)
Comment (satheeshravi) [inline lib/Backend/Lower.cpp]: Still this is kinda multiple implicit assumptions about the operands - instead if you rely on the slotNum, it will be more explicit and more straightforward about whether you need a fast path or not.

---
In reply to: [71741291](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71741291) [](ancestors = 71741291,71649040)

Comment (satheeshravi) [inline lib/Backend/Lower.cpp]: You can, if you are looking at the slotNum to figure out the number of arguments.

---
In reply to: [71743911](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71743911) [](ancestors = 71743911,71650995,71630471)

Comment (kunalspathak) [inline lib/Backend/Lower.cpp]: Didn't know that there is an API for `GetArgSlotNum` I can utilize. I will restructure the code.

---
In reply to: [71782791](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71782791) [](ancestors = 71782791,71743911,71650995,71630471)

File: lib/Backend/Lower.cpp (hunk 2 above)
Comment (kunalspathak) [inline lib/Backend/Lower.cpp]: > You can still call GetImmediateValue on the addr opnd,

`GetImmediateValue` returns intptr_t, which for `OpndKindAddr` just returns the address.

---
In reply to: [71782041](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71782041) [](ancestors = 71782041,71742831,71649822)

File: lib/Backend/Lower.cpp [hunk 9]
```diff
@@ -3833,16 +3835,18 @@ Lowerer::GenerateProfiledNewScObjArrayFastPath(IR::Instr *instr, Js::ArrayCallSi
     Func * func = this->m_func;
     IR::LabelInstr * helperLabel = IR::LabelInstr::New(Js::OpCode::Label, func, true);
     uint32 size = length;
+    bool containMissingItems = (size != 0);
     bool isZeroed;
```
Comment (satheeshravi) [inline lib/Backend/Lower.cpp]: >isZeroed [](start = 9, length = 8)

please initialize this one, though you didn't introduce this :)

**Comment by satheeshravi** [comment]
:shipit:

**Comment by kunalspathak** [comment]
@satheeshravi , Thanks for reviewing.
